src/base/utils/tar.cpp

```diff
@@ -0,0 +1,298 @@
+ * exception statement from your version.
+ */
+
+#include "tar.h"
+
+#include <QByteArray>
+#include <QCoreApplication>
+#include <QHash>
+#include <QString>
+
+#include "base/logger.h"
+#include "base/utils/bytearray.h"
+#include "base/utils/fs.h"
+
+namespace
+{
+    const int TAR_HEADER_SIZE = 512;
+
+    const int TAR_FILENAME_OFFSET = 0;
+    const int TAR_FILESIZE_OFFSET = 124;
+    const int TAR_CHECKSUM_OFFSET = 148;
+    const int TAR_FILETYPE_OFFSET = 156;
+    const int TAR_FILENAME_PREFIX_OFFSET = 345;
+
+    const int TAR_FILENAME_SIZE = 100;
+    const int TAR_FILESIZE_SIZE = 12;
+    const int TAR_CHECKSUM_SIZE = 8;
+    const int TAR_FILENAME_PREFIX_SIZE = 155;
+
+    class TarEntry
+    {
+    public:
+        enum Types : unsigned char
+        {
+            REG_FILE_NULL = '\0',
+            REG_FILE_ASCII = '0',
+            // We don't care about the other file types
+
+            INVALID = std::numeric_limits<unsigned char>::max()
+        };
+
+        TarEntry() = default;
+
+        explicit TarEntry(const QByteArray &header)
+            : m_header(header)
+        {
+            if (m_header.size() != TAR_HEADER_SIZE)
+                return;
+
+            // Invalid header: The first header field represents the filename.
+            // If the 1st byte is null then we can't extract a name. So the whole header becomes invalid to us.
+            m_valid = !m_header.startsWith('\0');
+            verifyAndCheckNull();
+        }
+
+        QString name() const
+        {
+            if (!isValid())
+                return {};
+
+            const QByteArray filename = Utils::ByteArray::midView(m_header, TAR_FILENAME_OFFSET, TAR_FILENAME_SIZE);
+            const int filenameLen = strlen(filename);
+            if (filenameLen == 0)
+                return {};
+
+            const QByteArray filenamePrefix = Utils::ByteArray::midView(m_header, TAR_FILENAME_PREFIX_OFFSET, TAR_FILENAME_PREFIX_SIZE);
+            const int filenamePrefixLen = strlen(filenamePrefix);
+
+            const QString finalFilename = (filenamePrefixLen > 0) ?
+                                        QString::fromUtf8(filenamePrefix.constData(), filenamePrefixLen)
+                                              + '/' + QString::fromUtf8(filename.constData(), filenameLen)
+                                      : QString::fromUtf8(filename.constData(), filenameLen);
+
+            return Utils::Fs::toValidFileSystemName(finalFilename, true);
+        }
+
+        bool isNull() const { return m_null; }
+        bool isValid() const { return m_valid; }
+        bool isVerified() const { return m_verified; }
+        Types type() const { return isValid() ? static_cast<Types>(m_header.at(TAR_FILETYPE_OFFSET)) : INVALID; }
+        int size() const { return isValid() ? octalToDec(Utils::ByteArray::midView(m_header, TAR_FILESIZE_OFFSET, TAR_FILESIZE_SIZE)) : 0; }
+
+    private:
+        int octalToDec(const QByteArray &data) const
```

Comment:
This can be extracted to anonymous namespace.
And `int strlen(const QByteArray &str)` too.